[addons] add interface helper class to translate "C" to "CPP"

AlwinEsch opened the pull request:
## Description
This is currently not used, thats why also no Version change is included.
On coming changes with Game and PVR addon system becomes this necessary.

Separately needed at two places and there to keep smaller requests. It
is also usable by others, regardless of the addon types and in the future.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
Related to:
- https://github.com/xbmc/xbmc/pull/16485
- https://github.com/xbmc/xbmc/pull/16453
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +294,94 @@ class CSettingValue
+
+protected:
+  bool m_owner = false;
+  C_STRUCT* m_cStructure;
```
... = nullptr;

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +294,94 @@ class CSettingValue
+ *
+ * ~~~~~~~~~~~~~
+ *
+ * The class automatically works with a "C" = "CPP" or "CPP" = "C".
```
Sorry, I do not get what this means.

AlwinEsch replied:
Removed and added example instead.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +294,94 @@ class CSettingValue
+  }
+
+  const CPP_CLASS& operator =(const CPP_CLASS& right)
+  {
```
```c++
if (m_cStructure && m_owner)
{
  delete m_cStructure;
  m_cStructure = nullptr;
}
```

before changing m_owner and assigning m_cStructure?

AlwinEsch replied:
Also here to set with nullptr bring hidden errors / crashes.

Only this `m_owner` in the class can really make private instead of protected.

AlwinEsch replied:
Ah you mean? Yes that could be more safe to prevent memleaks.
```cpp
  const CPP_CLASS& operator =(const CPP_CLASS& right)
  {
    if (m_cStructure && m_owner)
      delete m_cStructure;
    m_owner = true;
    m_cStructure = new C_STRUCT(*right.m_cStructure);
    return *this;
  }
```

bugwelle replied:
The check in the if condition for `m_cStructure` is also not necessary. Calling `delete` with a `nullptr` is well defined since C++14:

> If expression evaluates to a null pointer value, no destructors are called, and the deallocation function is not called.  
> https://en.cppreference.com/w/cpp/language/delete

ksooo replied:
Right, but why should I call delete if I know it will do nothing?

ksooo replied:
> Calling delete with a nullptr is well defined since C++14:

 Kodi is c++ 11

AlwinEsch replied:
No more since yesterday https://github.com/xbmc/xbmc/pull/13034

ksooo replied:
Nice to get to know about such an important change this side way. :-/

AlwinEsch replied:
You was linked as reviewer there.

ksooo replied:
Whoever requested my review there seems not to know that I almost know nothing about cmake. ;-)

bugwelle replied:
> Right, but why should I call delete if I know it will do nothing?

Right, but it's a superfluous check because "delete" already checks if the ptr is null. Removing it would (in my opinion) make the code more readable. Just my opinion, though :-)
But that's up to the developer.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +294,94 @@ class CSettingValue
+  const CPP_CLASS& operator =(const CPP_CLASS& right)
+  {
+    m_owner = true;
+    m_cStructure = new C_STRUCT(*right.m_cStructure);
```
```c++
if ((*right.m_cStructure)
  m_cStructure = new C_STRUCT(*right.m_cStructure);
```

AlwinEsch replied:
Having it with the "if" would rather bring in hidden errors.

m_cStructure would be NULL and would be called later by the child class.

Maybe for this case:
```cpp
  const CPP_CLASS& operator =(const CPP_CLASS& right)
  {
    m_owner = true;
    if (*right.m_cStructure)
      m_cStructure = new C_STRUCT(*right.m_cStructure);
    else
    {
      kodi::Log(ADDON_LOG_FATAL, "CStructHdl::%s: called with nullptr!", __func__);
      m_cStructure = new C_STRUCT;
    }
    return *this;
  }
```
But then also rather hidden and could bring confusion.

Thinking it would be safest to do so, and you would discover something in development:
```cpp
  const CPP_CLASS& operator =(const CPP_CLASS& right)
  {
    assert(*right.m_cStructure)
    m_owner = true;
    m_cStructure = new C_STRUCT(*right.m_cStructure);
    return *this;
  }
```

ksooo replied:
Okay. If it is an error that `right` has no `m_cStructure`, then an `assert` is just fine.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +294,94 @@ class CSettingValue
+  {
+    m_owner = true;
+    m_cStructure = new C_STRUCT(*right);
+    return *this;
```
Same comments as for the other `operator =`

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +294,94 @@ class CSettingValue
+
+  operator C_STRUCT*() { return m_cStructure; }
+
+protected:
```
Why protected, not private? Do you expect derived classes?

AlwinEsch replied:
With private, can child not access the structure, see: https://github.com/xbmc/xbmc/blob/caf5e7e3c39ee8337f3cec182b2748031e72d134/xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/PVRClient.h#L588

ksooo replied:
child means derived class?

AlwinEsch replied:
Exact, as on the example link.

ksooo replied:
Okay, then.

ksooo reviewed (commented):
Some questions/remarks.

ksooo reviewed (approved):
LGTM now.

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -8,6 +8,7 @@
 
 #pragma once
 
+#include <assert.h>     /* assert */
```
> Do not use whitespace to vertically align [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+ * // Give "data" to Kodi
+ * ~~~~~~~~~~~~~
+ */
+template <class CPP_CLASS, typename C_STRUCT>
```
no space before `<`

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+{
+public:
+  CStructHdl()
+    : m_cStructure(new C_STRUCT),
```
remove comma in this line

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+public:
+  CStructHdl()
+    : m_cStructure(new C_STRUCT),
+      m_owner(true)
```
add comma in this line

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+
+  CStructHdl(const CPP_CLASS& cppClass)
+    : m_cStructure(new C_STRUCT(*cppClass.m_cStructure)),
+      m_owner(true)
```
same

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+
+  CStructHdl(const C_STRUCT* cStructure)
+    : m_cStructure(new C_STRUCT({*cStructure})),
+      m_owner(true)
```
same

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+
+  CStructHdl(C_STRUCT* cStructure)
+    : m_owner(false),
+      m_cStructure(cStructure)
```
same

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+    assert(cStructure);
+  }
+
+  const CPP_CLASS& operator =(const CPP_CLASS& right)
```
remove space between `operator` and `=`

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -294,6 +295,110 @@ class CSettingValue
+    return *this;
+  }
+
+  const CPP_CLASS& operator =(const C_STRUCT& right)
```
remove space between `operator` and `=`

AlwinEsch committed:
[addons] add interface helper class to translate "C" to "CPP"

This is currently not used, thats why also no Version change is included.
On coming changes with Game and PVR addon system becomes this necessary.

Separately needed at two places and there to keep smaller requests. It
is also usable by others, regardless of the addon types and in the future.
